arraycontext/container/traversal.py

```diff
@@ -629,12 +632,36 @@ def _flatten(subary: ArrayOrContainerT) -> None:
             for _, isubary in iterable:
                 _flatten(isubary)
 
-    _flatten(ary)
+    def _flatten_without_leaf_class(subary: ArrayOrContainerT) -> Any:
+        nonlocal result, common_dtype
+        result = []
+        common_dtype = None
+
+        _flatten(ary)
+
+        if len(result) == 1:
+            return result[0]
+        else:
+            return actx.np.concatenate(result)
+
+    def _flatten_with_leaf_class(subary: ArrayOrContainerT) -> Any:
+        if type(subary) is leaf_class:
+            return _flatten_without_leaf_class(subary)
+
+        try:
+            iterable = serialize_container(subary)
+        except NotAnArrayContainerError:
+            return subary
```

Comment:
Should this do a `np.ravel(subary)` so that all the arrays are 1d?

Reply:
We're "above" the leaf class here, so I'd say it's consistent to leave the arrays alone.